src/services/services.ts

```diff
@@ -6259,6 +6373,187 @@ namespace ts {
 
                     return result[index];
                 }
+
+                function addReferenceToRelatedSymbol(node: Node, relatedSymbol: Symbol) {
+                    const references = getReferencedSymbol(relatedSymbol).references;
+                    if (implementations) {
+                        getImplementationReferenceEntryForNode(node, references);
+                    }
+                    else {
+                        references.push(getReferenceEntryFromNode(node));
+                    }
+                }
+            }
+
+            function getImplementationReferenceEntryForNode(refNode: Node, result: ReferenceEntry[]): void {
+                // Check if we found a function/propertyAssignment/method with an implementation or initializer
+                if (isNameOfImplementation(<Identifier>refNode)) {
+                    result.push(getReferenceEntryFromNode(refNode.parent));
+                }
+                else if (refNode.kind === SyntaxKind.Identifier) {
+                    if (refNode.parent.kind === SyntaxKind.ShorthandPropertyAssignment) {
+                        // Go ahead and dereference the shorthand assignment by going to its definition
+                        getReferenceEntryForShorthandPropertyAssignment(refNode, typeChecker, result);
+                    }
+
+                    // Check if the node is within an extends or implements clause
+                    const containingHeritageClause = getContainingClassHeritageClause(refNode);
+                    if (containingHeritageClause) {
+                        result.push(getReferenceEntryFromNode(containingHeritageClause.parent));
+                        return;
+                    }
+
+                    // If we got a type reference, try and see if the reference applies to any expressions that can implement an interface
+                    const containingTypeReference = getContainingTypeReference(refNode);
+                    if (containingTypeReference) {
+                        const parent = containingTypeReference.parent;
+                        if (isVariableLike(parent) && parent.type === containingTypeReference && parent.initializer && isImplementationExpression(parent.initializer)) {
+                            maybeAdd(getReferenceEntryFromNode(parent.initializer));
+                        }
+                        else if (isFunctionLike(parent) && parent.type === containingTypeReference && parent.body && parent.body.kind === SyntaxKind.Block) {

```

Comment:
If the body isn't a block, can't you still check for an implementation right there? E.g.:
```ts
interface I { x: number }
(x: number): I => ({ x })
```